# ДЗ 1.

### Задача 1

Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).



In [2]:
import numpy as np
from math import factorial


In [3]:
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

Пусть X - дискретная случайная величина, равная числу правильных ответов. Она распределена по биномиальному закону с параметрами n = 5 (число вопросов), p = 1/4 (вероятность угадать ответ) 

In [4]:
n = 5
p = 1/4
q = 1 - p
m = n * p
d = n * p * q
print(f'Математическое ожидание M(X) равно {m}')
print(f'Дисперсия D(X) равна  {d}')

Математическое ожидание M(X) равно 1.25
Дисперсия D(X) равна  0.9375


Воспользуемся формулой Бернулли для расчета вероятности
$$P(X = k) = C_n^k p^k (1 - p)^{n - k}$$

In [5]:
k = 0
p0 = combinations(n, k) * (p**k) * (q**(n-k))
print(p0)

0.2373046875


In [6]:
k = 1
p1 = combinations(n, k) * (p**k) * (q**(n-k))
print(p1)

0.3955078125


In [7]:
k = 2
p2 = combinations(n, k) * (p**k) * (q**(n-k))
print(p2)

0.263671875


In [8]:
k = 3
p3 = combinations(n, k) * (p**k) * (q**(n-k))
print(p3)

0.087890625


In [9]:
k = 4
p4 = combinations(n, k) * (p**k) * (q**(n-k))
print(p4)

0.0146484375


In [10]:
k = 5
p5 = combinations(n, k) * (p**k) * (q**(n-k))
print(p5)

0.0009765625


Закон распределения такой случайной величины:
<table border="3">
    <tr>
        <th>$x$</th>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
        <td>5</td>
    </tr>
    <tr>
        <th>$P(X = x)$</th>
        <td>0.2373</td>
        <td>0.3955</td>
        <td>0.2636</td>
        <td>0.0879</td>
        <td>0.0146</td>
        <td>0.0009</td>
    </tr>
</table>    

### Задача 2

Вероятность попадания в цель при одном выстреле равна 0.01. Какова вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10?

Так как число попаданий не менее 5 и не более 10, будем считать вероятность для 5, 5, 7, 8, 8, 10 попаданий


In [11]:
p = 0.01
q = 1 - p
n = 200
k5 = 5
k6 = 6
k7 = 7
k8 = 8
k9 = 9
k10 = 10


In [12]:
p5 = combinations(n, k5) * (p**k5) * (q**(n-k5))
p6 = combinations(n, k6) * (p**k6) * (q**(n-k6))
p7 = combinations(n, k7) * (p**k7) * (q**(n-k7))
p8 = combinations(n, k8) * (p**k8) * (q**(n-k8))
p9 = combinations(n, k9) * (p**k9) * (q**(n-k9))
p10 = combinations(n, k10) * (p**k10) * (q**(n-k10))


Так как события несовместные, вероятности складываем:

In [13]:
p = p5 + p6 + p7 + p8 + p9 + p10
print(f'Вероятность равна {p}')

Вероятность равна 0.051739381860374606


### Задача 3

В условиях предыдущей задачи сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание?

Подсказка. Как посчитать P(X >= k) для величины X, имеющей распределение Пуассона? Здесь предстоит немножко покодить.

Вероятность попадания в цель p = 0.01
В условии указано что хотя бы 10 раз будет совершено попадание, значит k>=10
Вероятность P = 0.9
Сумма вероятностей случайной величины всегда равна 1. Значит, вероятность получить хотя бы 10 (т. е 10 и более попаданий) равна:
$$P(X >= 10) = 1 - P(X < 10)$$
Получается:
$$0.9 = 1 - P(X < 10)$$
$$P(X < 10) = 0.1$$

Как из данного равенства выразить n выстрелов я не поняла, получается очень сложное уравнение, где и n в степени 10, и n является показателем степени (с минусом). Можно попробовать перебрать много вариантов в цикле. 

In [32]:
def poisson_proba(k, lambda_):
    
    return (lambda_ ** k) * (np.exp(-lambda_)) / np.math.factorial(k)


p = 0.01
list_n = []
list_r = []
for n in range (2000):
    lambda_ = n * p
    r = 1 - sum(poisson_proba(k=i, lambda_=lambda_) for i in range(10))
    if 0.9 <= r <= 0.901:
        list_n.append(n)
        list_r.append(r)
        
print(list_n)
print(list_r)


[1421, 1422]
[0.9001760722984115, 0.9006140609380022]


In [33]:
n = min(list_n)
print(f' {n} выстрел нужно сделать чтобы было хотя бы 10 попаданий')

 1421 выстрел нужно сделать чтобы было хотя бы 10 попаданий


In [47]:
#Проверка:
lambda_ = 1421 * 0.01
1 - sum(poisson_proba(k=i, lambda_=lambda_) for i in range(10))


0.9001760722984115